etc/profile-m-z/telegram.profile

```diff
@@ -41,7 +44,7 @@ seccomp.block-secondary
 shell none
 
 disable-mnt
-private-bin telegram,Telegram,telegram-desktop
+private-bin bash,sh,telegram,Telegram,telegram-desktop,xdg-open
```

Comment:
Why bash?

xdg-open is a shell-script, consider `sed,xdg-mime,which,mimeopen,grep,egrep,printf,cut,uname,dbus-send,xprop,dirname,cat,…`, `exo-open,gio,gvfs-open,mate-open,enlightenment_open,gnome-open,dde-open,kde-open,kde-open*,kfmclient,cygstart,kde-config,gnome-default-applications-properties,…` or just drop private-bin.

Reply:
> Why bash?

I believe enabling `bash` is necessary because in Arch Linux, `/usr/bin/sh` has been symlink'ed to `/usr/bin/bash`. Some customized systems also use `bash` as `sh`. I personally frequently use `bash` as `sh` in production systems too because `bash` is so much more featureful than `sh`. In fact, `bash` has a `sh` compatible mode which is probably why I've never had any issue when replacing `sh` with `bash`. Finally, in this case, I tested only enabling `sh`, and confirmed that it didn't work.

Reply:
> I believe enabling bash is necessary because in Arch Linux, `/usr/bin/sh` has been symlink'ed to `/usr/bin/bash`

Do not believe, test :wink:.

private-bin follows symlinks IIRC and even if it don't there are other shells commonly used as `/bin/sh` (namely `dash` by Debian and Ubuntu(?)).

> Finally, in this case, I tested only enabling sh, and confirmed that it didn't work.

(Fedora Linux 35; `sh->bash`):

```console
$ firejail --noprofile --private-bin=sh,ls ls -l /usr/bin
bash
ls
sh -> /usr/bin/bash
```

I guess you had tested with only `sh` in `private-bin` and only `noblacklist ${PATH}/sh`. This does not work `blacklist ${PATH}/bash` will blacklist the binary used by `/usr/bin/sh`. FWIW, `blacklist` follows symlinks too.

Reply:
> xdg-open is a shell-script, consider …

*On my system* for xdg-open to work at all: `sh,xdg-open,grep,egrep,<A installed well known browser, e.g. firefox>`.
*On my system* for xdg-open to work correct: `sh,xdg-open,grep,egrep,xdg-mime,sed,tr,awk,cut,head,basename,which,readlink,<My browser, e.g. firefox>`
For the firefox start script: `...`

> or just drop private-bin.

yeah, I think this is the way to go.

Reply:
> I guess you had tested with only `sh` in `private-bin` and only `noblacklist ${PATH}/sh`

That's correct, I think tested with only `sh` in `private-bin` and I only `noblacklist` `sh`.

> there are other shells commonly used as /bin/sh (namely dash by Debian and Ubuntu(?))

I think `dash` is indeed an implementation of POSIX `sh` which is actually `sh`.

Reply:
@rusty-snake [on Dec 19](https://github.com/netblue30/firejail/pull/4783#discussion_r771977694):

> > I believe enabling bash is necessary because in Arch Linux, `/usr/bin/sh`
> > has been symlink'ed to `/usr/bin/bash`
> 
> Do not believe, test wink.
> 
> private-bin follows symlinks IIRC and even if it don't there are other shells
> commonly used as `/bin/sh` (namely `dash` by Debian and Ubuntu(?)).
> 
> > Finally, in this case, I tested only enabling sh, and confirmed that it didn't work.
> 
> (Fedora Linux 35; `sh->bash`):
> 
> ```
> $ firejail --noprofile --private-bin=sh,ls ls -l /usr/bin
> bash
> ls
> sh -> /usr/bin/bash
> ```

This just lists the files; /bin/sh could point to `foobar` and it would not
fail because of that.  On Artix it fails, but I'm not sure why:

```console
$ firejail --quiet --noprofile --private-bin=sh,ls ls -l /usr/bin
Cannot start application: No such file or directory
```

A better test would be to actually run /bin/sh:

```console
$ firejail --quiet --noprofile --private-bin=sh /bin/sh -c 'echo yes'
Cannot start application: No such file or directory
$ firejail --quiet --noprofile --private-bin=bash,sh /bin/sh -c 'echo yes'
yes
$ readlink /bin/sh
dash
```

So in my case it does require `private-bin bash`, even though /bin/sh is
/bin/dash.

IIRC this is because firejail hardcodes bash as the default shell.

Related:

* <https://github.com/netblue30/firejail/issues/3434#issuecomment-923552299>

Maybe firejail could just use `readlink /bin/sh` once on startup and consider
that the default shell.

Reply:
> This just lists the files;

@kmk3 it clearly shows that the file `/bin/sh` points to is copied/bind-mounted (what does it do?) if you only list `sh` but not `bash` (assuming `/bin/sh` points to somewhere in `{/usr,}/bin`).

Reply:
> IIRC this is because firejail hardcodes bash as the default shell.

It don't, read the issues title in your comment:

> - Default shell is guessed from $SHELL, despite manpage specifying /bin/bash

---

> On Artix it fails, but I'm not sure why:

If you get your right (sh->dash; `getent passwd $USER | cut -d: -f7`: bash), you try to start a program (bash) which isn't inside the sandbox.

IMHO `--shell=none` should be the default (if you specify a program).

Reply:
@rusty-snake [on Dec 19](https://github.com/netblue30/firejail/pull/4783#discussion_r771991938):

> > This just lists the files;
> 
> @kmk3 it clearly shows that the file `/bin/sh` points to is
> copied/bind-mounted (what does it do?) if you only list `sh` but not `bash`
> (assuming `/bin/sh` points to somewhere in `{/usr,}/bin`).

I don't understand what you mean.  A symlink points to a (text) path; it
doesn't actually point to an inode.  For example, if /bin/sh points to just
"dash" and you copy the symlink with to /tmp, the copy will not work:

```console
$ readlink /bin/sh
dash
$ /bin/sh -c 'echo yes'
yes
$ cd /tmp
$ cp -P /bin/sh .
$ readlink ./sh
dash
$ ./sh -c 'echo yes'
bash: ./sh: No such file or directory
$ echo 'echo hello world' >dash
$ chmod +x dash
$ ./dash
hello world
$ ./sh
hello world
```

Do you mean that sh points to just "bash" outside of firejail but inside of it
it points to "/usr/bin/bash"?  That might mean that firejail hardcodes its own
/bin/sh symlink inside the sandbox, but whether /bin/bash is bind-mounted or
not is unclear from your example.

Reply:
What I want to tell is that `private-bin=sh` will copy/bind-mount `sh` *and* the program it points to (e.g. bash or dash). No matter if you list `bash/dash` or not.

I don't care about `/bin/sh` pointing to somewhere else (e.g. `/mnt/extra-program/mysh`).

Reply:
I just tried removing `bash` from `private-bin`, and it worked (the link could be opened from Telegram). @kmk3 do you still have problem if removing `bash` from `private-bin`?

Reply:
(Continued on #4790)